### PR TITLE
BUG: first("2M") returning incorrect results

### LINKED ISSUE #29623: BUG: incorrect output of first('1M') in case if first index is the last day of the month
#### Code Sample

```python
x = pd.Series(1, index=pd.bdate_range('2010-03-31', periods=100))
print(x.first('1M'))  # Returns March and April
print(x.first('2M'))  # Returns March, April and May
# etc...
```
#### Problem description

In case if the first index of the series falls on the last day of the month, `Series.first('1M')` returns two first months (this one day + the next month).

If the first value is not on the last day of the month - result is correct:

```python
x = pd.Series(1, index=pd.bdate_range('2010-03-30', periods=100))
print(x.first('1M'))  # Returns only March
print(x.first('2M'))  # Returns March and April
# etc...
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.1.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.7.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : en_US.UTF-8
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.3
numpy            : 1.17.2
pytz             : 2019.3
dateutil         : 2.8.0
pip              : 19.3.1
setuptools       : 41.4.0
Cython           : None
pytest           : 5.2.1
hypothesis       : None
sphinx           : 2.2.0
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.4.1
html5lib         : None
pymysql          : 0.9.3
psycopg2         : None
jinja2           : 2.10.1
IPython          : 7.8.0
pandas_datareader: None
bs4              : 4.6.3
bottleneck       : None
fastparquet      : 0.1.6
gcsfs            : None
lxml.etree       : 4.4.1
matplotlib       : 3.0.1
numexpr          : 2.7.0
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 0.12.1
pytables         : None
s3fs             : 0.3.4
scipy            : 1.3.1
sqlalchemy       : 1.3.8
tables           : 3.4.4
xarray           : None
xlrd             : 1.1.0
xlwt             : None
xlsxwriter       : None

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #29623
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
 
This would fix #29623 for offsets with n > 1
cc @simonjayhawkins

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
Thanks @phofl for the quick response.

I'm not sure where this leaves us w.r.t. backports. whether we should backport #38331 (i.e. reopen #38436) and then backport this as well.

or let these changes sit in master for a while (and move release note added in #38331 to 1.3 whatsnew)

**Comment by phofl** [comment]
If we decide to backport, we could cherry pick this one on top of the other Commit. But I am not quite sure either if we should backport.

**Comment by jreback** [comment]
we can just revert the original and then merge to master

**Comment by phofl** [comment]
Oh right, that would work. Should I open a Revert PR?

**Comment by simonjayhawkins** [comment]
#38448

**Comment by phofl** [comment]
Ok, thx. Will adjust this one after the other is merged

**Comment by simonjayhawkins** [comment]
after the revert is merged, will reopen issue and change milestone. so can update the OP here to close issue again.

**Comment by phofl** [comment]
Done. This fixes both cases now.

File: doc/source/whatsnew/v1.2.0.rst [hunk 1]
```diff
@@ -607,6 +607,7 @@ Datetimelike
 - Bug in :meth:`Series.isin` with ``datetime64[ns]`` dtype and :meth:`.DatetimeIndex.isin` failing to consider timezone-aware and timezone-naive datetimes as always different (:issue:`35728`)
 - Bug in :meth:`Series.isin` with ``PeriodDtype`` dtype and :meth:`PeriodIndex.isin` failing to consider arguments with different ``PeriodDtype`` as always different (:issue:`37528`)
 - Bug in :class:`Period` constructor now correctly handles nanoseconds in the ``value`` argument (:issue:`34621` and :issue:`17053`)
+- Bug in :meth:`DataFrame.first` and :meth:`Series.first` returning two months for offset one month when first day is last calendar day (:issue:`29623`)
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.2.0.rst:610]: so we have three options

1. include in 1.2
2. change to 1.3
3. hold-off merging for a week or so, and then include in 1.2.1

i'm ok with 2 or 3, not so keen on 1.

Comment (simonjayhawkins) [reply doc/source/whatsnew/v1.2.0.rst:610]: > 3\. hold-off merging for a week or so

actually, no harm in adding a 1.2.1 whatsnew now.

Comment (phofl) [reply doc/source/whatsnew/v1.2.0.rst:610]: I am fine with either one of them

Comment (jreback) [reply doc/source/whatsnew/v1.2.0.rst:610]: ok let's push on 1.3

**Review by jreback — CHANGES_REQUESTED** [review]
changes look fine, let' sput on 1.3

File: doc/source/whatsnew/v1.2.0.rst (hunk 1 above)
Comment (phofl) [reply doc/source/whatsnew/v1.2.0.rst]: Done

**Review by jreback — APPROVED** [review]

File: pandas/core/generic.py [hunk 2]
```diff
@@ -8420,7 +8420,12 @@ def first(self: FrameOrSeries, offset) -> FrameOrSeries:
             return self
 
         offset = to_offset(offset)
-        end_date = end = self.index[0] + offset
+        if not isinstance(offset, Tick) and offset.is_on_offset(self.index[0]):
+            # GH#29623 if first value is end of period, remove offset with n = 1
+            #  before adding the real offset
+            end_date = end = self.index[0] - to_offset(offset.name) + offset
```
Comment (jbrockmendel) [inline pandas/core/generic.py:8426]: `offset.base` instead of `to_offset(offset.name)`?

Comment (phofl) [reply pandas/core/generic.py]: Better:) thx

**Comment by jreback** [comment]
thanks @phofl

**Pull request merged**